### ISSUE TITLE
link fields for in page navigation #go-to-here

### ISSUE THREAD

**Issue by serundeputy** [backdrop/backdrop-issues#1308]
- Add a link field to a content type say `Page`
- uncheck `Validate URL` in the field settings
- add an instance of the link field to `Page` content type
- add a `Page`
- in the link field
  - put some `Title`
  - in the `URL` text box put an internal hash `#go-to-here`

the link field is created, but the `hash` is encoded as `%23` resulting in link being 404 instead of in page nav.

Can we make in page anchors possible with link?

Reference:
http://www.w3.org/TR/html401/struct/links.html

**Comment by klonos** [comment]
I sorta like the idea, but I currently see no use for it. If I want nav links in a page, then I want them within the actual content/text. So I want nav link support in the rich text editor.

Say we do have a content type with a body and a link field and we do add a nav link in the link field. What would be the use case of having the user go to that field? I don't get it.

PS: maybe I'm totally stupid, but I was under the impression that you could use ```url-of-a-page#id-of-an-element``` in order to jump to that portion of the page. So perhaps what we should do is be auto-adding for example ```id="title"```, ```id="body"``` etc to all fields, so that people can create nav links that point to them.

**Comment by serundeputy** [comment]
the use case is this:

a user is going to make arbitrary number of fields on a page: `Unlimited` `field_my_awesome_content`
and they want `in-page-nav` to go to each of the fields.

So, we could have an unlimited link field to allow them to create the unlimited in page nav to correspond to each `field_my_awesome_content` that they create.

I can do this with text fields and `tpl`'s, but not with link fields as it encodes the `#`

>url-of-a-page#id-of-an-element

yes, that definitely works, but not with `link` fields as you can not put the `#`

**Comment by klonos** [comment]
Yes, I understand the issue. My doubt is this...

If you have unlimited link fields and you set them to ```#nav-whatever``` (once we sort out the ```#``` issue), they will point to where? ...the content fields? ...how?

The only way I see this being possible is with some sort of grouping (FieldGroup or Field Collection in core?) of say two fields, one of type link and one text. Then, you can add multiple/unlimited such pairs of field groups/collections where the link serves as a header/title for the corresponding text field. Then, you can refer to the various sections of that page from elsewhere by pointing to the headers

Unless you are talking about creating a separate page consisted of multiple/unlimited content fields and then another page with same number of link fields (acting as an index page for the page with the actual content)?

**Comment by serundeputy** [comment]
>The only way I see this being possible is with some sort of grouping (FieldGroup or Field Collection in core?) of say two fields, one of type link and one text.

yes; you've got it; they don't have to be in core, but for it to be possible; with `link` rather than my own fields it would need to allow `#`

**Comment by klonos** [comment]
Yes, then this issue here is about sorting the issue where link fields do not allow ```#```. Right?

**Comment by serundeputy** [comment]
>Yes, then this issue here is about sorting the issue where link fields do not allow #. Right?

yessir;

**Comment by klonos** [comment]
...because I thought it was about allowing fields to be assigned an id (so that they can be linked to).

**Comment by olafgrabienski** [comment]
I need to add internal anchor links using the link field as well. My use case: links to different sections of internal pages, e.g. `about-us#history`, `about-us#team` and so on. Which isn't possible because - as @serundeputy said in the issue description - the `#` gets encoded to `%23`.

Linking to anchors is perfectly possible if the links are *external* but it doesn't work for *internal* links:

- `http://example.com/about-us#history` works
`about-us#history` doesn't work

It seems that URL validation and encoding are different for both link types (in the link module, validation is beginning around line 1022). I haven't been able to find out more so far. @serundeputy Are you able to explain the exact difference in validation and/or encoding?

**Comment by yorkshire-pudding** [comment]
I think this happens in `function link_validate_url`
https://github.com/backdrop/backdrop/blob/1.x/core/modules/link/link.module#L1057-L1119

specifically, in this bit https://github.com/backdrop/backdrop/blob/0cd444f45c00eef18bf4ca70ac49ddc38116d4e3/core/modules/link/link.module#L1079-L1090

```php
  // Anchors must also be encoded, ASCII-only.
  $anchor = '(#[a-z0-9_\-\.~+%=&,$\'():;*@\[\]\/\?]*)';


  // The rest of the path for a standard URL.
  $end = $directories . '?' . $query . '?' . $anchor . '?$/i';


  // Add the end to the external pattern.
  $external_pattern .= $end;


  // Pattern specific to internal links.
  $internal_pattern = '/^([a-z0-9_\-+\[\] ]+)' . $end;
  $internal_pattern_file = '/^([a-z0-9_\-+\[\]\. \/\(\)][a-z0-9_\-+\[\]\. \(\)][a-z0-9_\-+\[\]\. \/\(\)]+)$/i';
```

I can't work out why it works for external but not internal.  I'm afraid pattern matching strings are still an enigma to me.

**Comment by yorkshire-pudding** [comment]
I'm not sure it is happening there after all. Looking in the database, it is stored intact.
![image](https://user-images.githubusercontent.com/20357797/168805343-d28a41c9-86d1-4230-b25b-abc544862110.png)

Could it be in the theme function? Or somewhere else that it turns the `#` into `%23`?

**Comment by indigoxela** [comment]
> Could it be in the theme function?

I guess so, yes, in theme_link_formatter_link_default(). It's caused by `l()`, which is supposed to urlencode properly - and it does. If one provides a fragment (`#something`), l() and url() expect that to go to the `$options` parameter.

So this all seems to work as expected, but obviously not as desired here. :wink:

One option could be to override theme_link_formatter_link_default in a custom module or theme. And then parse the url, extract any fragment and provide it in the $options array. Didn't test that, though.

**Comment by indigoxela** [comment]
Another option: if the absolute url works, why not using that instead: `[current-page:url:absolute]#test-block`?

**Comment by argiepiano** [comment]
Or perhaps the detection of fragments (and queries) can be done in _link_sanitize(), as done in this [Drupal 7 fix](https://www.drupal.org/project/link/issues/2654246) which I think did not make it into B. The functions differ somewhat, so this has to be looked into more deeply.

**Comment by yorkshire-pudding** [comment]
> Another option: if the absolute url works, why not using that instead: `[current-page:url:absolute]#test-block`?

Great idea - that does the trick for my use case.  It forces it to treat it as external so lets it through

**Comment by argiepiano** [comment]
> So I'm not sure if/how this would be supported without a checkbox or something that specifies the link is a fragment and sends it appropriately...

It seems to me that the solution I pointed out above (Drupal fix) takes care of this detection without adding things to the form

**Comment by argiepiano** [comment]
Though looking at it again it may be the case that the patch only takes care of  anchors or queries when they are provided by themselves and alone in the field... still, I think the approach of detecting the fragments early on before reaching the theme function is best. The theme function should do as little processing as possible

**Comment by BWPanda** [comment]
Yeah, the D7 patch essentially does what you ended up doing manually: if the link starts with `#`, add the full URL of the current page before it.

**Comment by yorkshire-pudding** [comment]
[1308.patch.txt](https://github.com/backdrop/backdrop-issues/files/8708714/1308.patch.txt)

Aye - this is the patch (modified to fit) and it doesn't fix the rendering

**Comment by BWPanda** [comment]
Related @indigoxela? https://github.com/backdrop/backdrop-issues/issues/1159

**Comment by indigoxela** [comment]
> Related?

Hm, not sure. It seems like #1159 tries to make things worse by allowing more space for human/machine misunderstandings.  (More ways to input "urls", that need to be considered in validation and rendering.)
